flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/file/PartitionFileIndexImpl.java

```diff
@@ -0,0 +1,159 @@
+            subpartitionRegions.add(new ArrayList<>());
+            subpartitionReaderRegionIndexes.add(new HashMap<>());
+        }
+    }
+
+    @Override
+    public Optional<Region> getNextRegion(
+            int subpartitionId, NettyConnectionId nettyServiceWriterId) {
+        synchronized (lock) {
+            if (isReleased) {
+                return Optional.empty();
+            }
+
+            int currentRegionIndex =
+                    subpartitionReaderRegionIndexes
+                            .get(subpartitionId)
+                            .getOrDefault(nettyServiceWriterId, 0);
+            List<Region> currentRegions = subpartitionRegions.get(subpartitionId);
+            if (currentRegionIndex < currentRegions.size()) {
+                Region region = currentRegions.get(currentRegionIndex);
+                ++currentRegionIndex;
+                subpartitionReaderRegionIndexes
+                        .get(subpartitionId)
+                        .put(nettyServiceWriterId, currentRegionIndex);
+                return Optional.of(region);
+            }
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public void addRegionForBuffers(List<SpilledBuffer> spilledBuffers) {
+        if (spilledBuffers.isEmpty()) {
+            return;
+        }
+
+        Map<Integer, List<Region>> convertedRegions = convertToRegions(spilledBuffers);
+        synchronized (lock) {
+            convertedRegions.forEach(
+                    (subpartition, regions) ->
+                            subpartitionRegions.get(subpartition).addAll(regions));
+        }
+    }
+
+    @Override
+    public void release() {
+        synchronized (lock) {
+            subpartitionRegions.clear();
+            subpartitionReaderRegionIndexes.clear();
+            isReleased = true;
+        }
+    }
+
+    // ------------------------------------------------------------------------
+    //  Internal Methods
+    // ------------------------------------------------------------------------
+
+    private static Map<Integer, List<Region>> convertToRegions(List<SpilledBuffer> spilledBuffers) {
+        Map<Integer, List<Region>> subpartitionRegionMap = new HashMap<>();
+        Iterator<SpilledBuffer> iterator = spilledBuffers.iterator();
+        SpilledBuffer firstBufferInRegion = iterator.next();
+        SpilledBuffer lastBufferInRegion = firstBufferInRegion;
+
+        while (iterator.hasNext()) {
+            SpilledBuffer currentBuffer = iterator.next();
+            if (currentBuffer.getSubpartitionId() != firstBufferInRegion.getSubpartitionId()
+                    || currentBuffer.getBufferIndex() != lastBufferInRegion.getBufferIndex() + 1) {
+                // the current buffer belongs to a new region, close the previous region
+                addInternalRegionToMap(
+                        firstBufferInRegion, lastBufferInRegion, subpartitionRegionMap);
```

Comment:
It seems we require `spilledBuffers` to be continuous, which is never mentioned in the contract.